supabase/migrations/18_refresh_tokens.sql

```diff
@@ -0,0 +1,120 @@
+-- Refresh tokens used to generate access tokens
+create table refresh_tokens (
+  like internal._model including all,
+  user_id    uuid references auth.users(id) not null,
+  multi_use  boolean default false,
+  valid_for  interval not null,
+  uses       int default 0,
+  hash       text not null
+);
+
+create policy "Users can access only their own refreshed_tokens"
+  on refresh_tokens as permissive
+  using (user_id = auth.uid());
+
+grant select(id, created_at, detail, updated_at, user_id, multi_use, valid_for, uses) on refresh_tokens to authenticated;
+grant update(detail, valid_for, multi_use) on refresh_tokens to authenticated;
+grant delete on refresh_tokens to authenticated;
+
+-- Create a new refresh_token
+create function create_refresh_token(multi_use boolean, valid_for interval, detail text default null)
+returns json as $$
+declare
+  secret text;
+  refresh_token_row refresh_tokens;
+begin
+  secret = gen_random_uuid();
+
+  insert into refresh_tokens (detail, user_id, multi_use, valid_for, hash)
+  values (
+    detail,
+    auth_uid(),
+    multi_use,
+    valid_for,
+    crypt(secret, gen_salt('bf'))
+  ) returning * into refresh_token_row;
+
+  return json_build_object(
+    'id', refresh_token_row.id,
+    'secret', secret
+  );
+commit;
+end
+$$ language plpgsql volatile security definer;
+
+-- Returns the secret used for signing JWT tokens, with a default value for
+-- local env, taken from https://github.com/supabase/supabase-js/issues/25#issuecomment-1019935888
+create function internal.access_token_jwt_secret()
+returns text as $$
+
+  select coalesce(current_setting('app.settings.jwt_secret', true), 'super-secret-jwt-token-with-at-least-32-characters-long') limit 1
+
+$$ language sql stable security definer;
```

Comment:
nit: this shouldn't be security definer, as that allows an unprivileged caller to run with the privilege of the definer. It doesn't matter in practice, as this is `internal` schema.